packages/xl-ai/src/streamTool/StreamToolExecutor.ts

```diff
@@ -0,0 +1,179 @@
+  }
+
+  private createWriteStream() {
+    let lastParsedResult: Operation<T> | undefined;
+
+    const stream = new TransformStream<string | Operation<T>, Operation<T>>({
+      transform: (chunk, controller) => {
+        const operation =
+          typeof chunk === "string"
+            ? partialJsonToOperation(
+                chunk,
+                lastParsedResult?.isPossiblyPartial ?? false,
+                this.streamTools,
+              )
+            : chunk;
+        if (operation) {
+          // TODO: string operations have been validated, but object-based operations have not.
+          // To make this consistent, maybe we should extract the string parser to a separate transformer
+          lastParsedResult = operation;
+          controller.enqueue(operation);
+        }
+      },
+
+      flush: (controller) => {
+        // Check if the stream ended with a partial operation
+        if (lastParsedResult?.isPossiblyPartial) {
+          controller.error(new Error("stream ended with a partial operation"));
+        }
+      },
+    });
+
+    return stream;
+  }
+
+  private createReadableStream() {
+    // this is a bit hacky as it mixes async iterables and streams
+    // would be better to stick to streams
+    let currentStream: AsyncIterable<Operation<StreamTool<any>[]>> =
+      createAsyncIterableStream(this.stream.readable);
+    for (const tool of this.streamTools) {
+      currentStream = tool.execute(currentStream);
+    }
+
+    return asyncIterableToStream(currentStream);
+  }
+
+  /**
+   * Helper method to apply all operations to the editor if you're not interested in intermediate operations and results.
+   */
+  public async waitTillEnd() {
+    const iterable = createAsyncIterableStream(this.readable);
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    for await (const _result of iterable) {
+      // no op
+      // these will be operations without a matching StreamTool.
+      // (we probably want to allow a way to access and handle these, but for now we haven't run into this scenario yet)
+    }
+  }
+
+  /**
+   * Returns a WritableStream that can be used to write StreamToolCalls to the executor.
+   *
+   * The WriteableStream accepts JSON strings or Operation objects.
+   */
+  public get writable() {
+    return this.stream.writable;
+  }
+
+  /**
+   * Accepts an async iterable and writes each chunk to the internal stream.
+   *
+   * (alternative to writing to the writable stream using {@link writable})
+   */
+  async execute(source: AsyncIterable<string | Operation<T>>): Promise<void> {
+    const writer = this.writable.getWriter();
+    for await (const chunk of source) {
+      await writer.write(chunk);
+    }
+    await writer.close();
+  }
+
+  /**
+   * Accepts a single chunk and processes it using the same logic.
+   *
+   * (alternative to writing to the writable stream using {@link writable})
+   */
+  async executeOne(chunk: StreamToolCall<T>): Promise<void> {
```

Comment:
Should be noted that running this once, will close the stream, so you can't call it in a loop or anything (since the .execute method closes the writer once it is done).

imo, this is a bit unexpected, I'd expect to be able to apply multiple executeOne calls, but I see how this is overlapping with just a writer.write.